**1. (20 points) True or False.**
 - **(a) GridSearchCV and Optuna will always produce the same results.** False
 - **(b) Support vector machine is an ensemble learning technique.** False
 - **(c) If the classifier produce a 80% recall on test dataset (using default values for hyperparameters), there is no need to tune the classifier hyper-parameters.** False
 - **(d) RandomSearchCV and Optuna will always produce the same results.** False
 - **(e) Increasing the value of max depth in tree-based models prevents overfitting.** False
 - **(f) Scaling the input variables in tree-based models can speed up the hyper-parameter tuning process.** False
 - **(g) In GridSearchCV, the number of trials can be specified by the practitioner.** False
 - **(h) XGBoost always outperforms random forest.** False
 - **(i) XGBoost is faster than regular gradient boosting algorithms because it trains the the base-learners in parallel.** True
 - **(j) In order to avoid overfitting, it is recommended to tune the learning rate hyperparameter in *random forest* models.** False

**2. (6 points) In what scenarios, RandomSearchCV would be preferred instead of GridSearchCV? Be specific.**

RandomSearchCV may be preferred over GridSearchCV when the hyperparameter search space is large because GridSearchCV looks at all combinations. It would also be prefered when computational resources are limited. It would also be prefered if you aren't trying to find the best or most optimal combination but one that performs well enough.

**3. (6 points) How does stacking work? Be specific**

Stacking is an ensemble learning technique that combines the predictions of multiple base models to make a final prediction. The idea behind stacking is to combine the strengths of the different base models by allowing the meta-learner to learn how to weight the predictions of the base models to improve prediction power. By doing so, the meta-learner can potentially make better predictions than the base models can on their own.

**4. (6 points) If you have trained three different models on the exact same training dataset, and they all achieve a 93% F1-score on the exact same dataset. Is there any chance that you can combine these three models to get better results? If so, how? If not, why? Be specific.**


Yes, you could still combine the three models and get improved results. By using the likelihoods from each of the models, you can create an ensemble model that takes the likelihoods as input variables to predict the class. Even if the three models have achieved an F1-Score of 93%, it is still possible for an ensemble model to improve prediction power because it can find new patterns in the likelihoods and potentially get better results.

**5. (4 points) If your XGBoost model overfits the data (probably because the number of boosted trees is large), what hyper-parameter you need to tune? Be specific.**

You could tune a number of variables to reduce overfitting. For example, you'd likely want to lower the learning rate which would reduce the influence of each individual tree and could lead to a model that generalizes better to new data. So, the model is less sensitive to the large number of trees that was likely the cause.

Alternatively, a large number of trees can also cause overfitting on its own. When there are too many trees, the model becomes too complex and starts fitting the noise in the data, instead of underlying patterns. By decreasing the number of trees, the model becomes simpler and more generalizable, which can help reduce overfitting.

**6. (4 points) Which of the following algorithm is NOT an ensemble learning algorithm?**
 - **(a) support vector machine**
 - **(b) random forest**
 - **(c) AdaBoost**
 - **(d) XGBoost**
 - **(e) None of the above**

(a) SVM

**7. (6 points) How does XGBoost with random forest as base learner work? Be specific.**

XGBoost does not allow for Random Forest to be selected as an option for the 'booster' hyper-parameter. However, Random Forest can still be incorporated into an XGBoost model by adjusting the 'num_parallel_trees' hyper-parameter to a value greater than 1. Essentially, this modification allows the model to use multiple trees in parallel for each boosting round, similar to a Random Forest model.

**8. (4 points) How do we select the best hyper-parameter combination for a given model?**
 - **(a) Performance on the training dataset**
 - **(b) Performance on the validation dataset**
 - **(c) (a) and (b)**
 - **(d) None of the above**

(b) Performance on the validation dataset

**9. The below chart shows the performance of a random forest models (with different number of trees).**

**Using the above chart, answer parts (a)-(b).**

**(a) (4 points) If a data scientist wants to tune the number of trees based on the F1-score, how many trees would he select? Be specific.**

If the data scientist is tuning the number of trees based on the F1-score, they should select 400 trees because there is little to no prediction improvement in the train and test sets after increasing the number of trees past 400. Increasing beyond this point would result in increased computation time with little to no model improvement.

**(b) (6 points) What is the reason there is big difference in the F1-score performance between the Train and Test datasets? Be specific.**

The biggest reason why there would be a huge difference in the F1-score performance between the train and test datasets is that the model is likely overfitting the on the training data. The model fits the training data very well, but does not do a great job of generalizing on the test set. 

**10. (6 points) What is the difference between homogeneous and heterogeneous ensembles? Be specific.**

A heterogeneous ensemble method refers to an ensemble model that uses different algorithms for the base learners, whereas a homogeneous ensemble refers to an ensemble model that uses the same algorithm for all the base learners on different distributions of the training dataset. Homogeneous ensembles are typically used to improve the performance of a single type of model, while heterogeneous ensembles are used to combine the strengths of multiple models. 

**11. (4 points) Random forest is an example of**
 - **(a) Homogeneous ensemble**
 - **(b) Heterogeneous ensemble**
 - **(c) Bagging**
 - **(c) (a) and (b)**
 - **(d) (a) and (c)**
 - **(e) (b) and (c)**
 - **(f) None of the above**

(d) (a) and (c)

**12. (8 points) A junior data scientist is working on building a classification framework that can help a small local credit union to identify fraudulent transactions. The data scientist collects a very small dataset with 100 observations, from which 4 observations are labelled as fraudulent, to start building and tuning the model. As part of the tuning process, the data scientist considers using the GridSearch function from scikit-learn with cv = 5. Explain why this approach is a good idea or not. Be specific.**

This approach is not a good idea because GridSearchCV uses a stratified k-fold approach, which means that by setting cv = 5, it splits the 100 observation set into five data sets that are meant to contained a stratified sample of the class of interest (fraud). However, since there are only 4 fraud observations for 5 folds, one fold is not going to have a fraudulent observation. Ultimately, the approach is not a good idea.

**13. The typical hyper-parameters in a XGBoost model that data science practitioners tune are:**
 - **n estimators**
 - **learning rate**
 - **max depth**
 - **min child weight**
 - **subsample**
 - **colsample bytree**
 
**(a) (4 points) What is/are the hyper-parameters that may cause overfitting if we increase their values while holding the values of the hyper-parameters constant? Be specific.**

Increasing the values for learning rate, max depth, subsample, and colsample bytree while holding all other values constant may cause overfitting.

**(b) (4 points) What is/are the hyper-parameters that prevents overfitting if we increase their values while holding the values of the hyper-parameters constant? Be specific.**

Increasing the value for min child weight while holding all other values constant may prevent overfitting.

**14. (6 points) In tree-based models such as random forest, gradient boosting, and XGBoost, how does a data scientist optimize the number of trees in a manually fashion? Be specific.**

The number of trees can be tuned with the n_estimators hyper-parameter. This value can be optimized/tuned with a tuning method like RandomizedSearchCV, GridSearchCV, or Optuna. Alternatively, the data scientist could optimize this number by testing various numbers for n_trees and plotting the performance of the model on the train and test sets. This would allow the data scientist to examinine the plotted curve and determine the optimal number of trees that maximizes the performance of the model on the test set while minimizing overfitting to the training set.

**15. A Machine Learning Specialist is assigned to a team that is responsible for tuning an XGBoost model to ensure that it performs correctly on test data. However, when dealing with unknown data, this does not operate as planned. The following table summarizes the existing hyper-parameters:**

    XGBoost_params = {'n_estimators': 2000,
                      'max_depth': 30,
                      'min_child_weight': 3,
                      'subsample': 0.9,
                      'objective': 'reg:squarederror'}
                      
**(a) (3 points) What type of task the machine learning specialist is working on? Classification or regression? Be specific.**

The machine learning specialist is working on a regression task because the objective is set to "reg:squarederror". This is the default setting for onjective in XGBoost regression tasks.

**(b) (5 points) Given the existing XGBoost hyper-parameters, what may be the reason that the XGBoost model is not performing as expected? Be specific.**

Looking at the hyper-parameters for the XGBoost model, it may not be performing as expected because the n_estimators and max_depth values are likely causing the model to be overfit. A max depth of 30 is very large and is likely the main culprit in causing the modle to overfit. Though increasing the number of trees can improve the result, 2000 may be also be too many to consider for their model. Similarly, decreasing the max depth will  likely decrease the likelihood that the model is overfit. Overall, they should consider changes to n_estimators and max_depth to reduce overfitting on the training data and hopefully see some increased model performance.

**16. (8 points) How can overfitting and underfitting be addressed through hyper-parameter tuning in random forest? That is, what hyper-parameters would you tune to address overfitting and underfitting in random forest? Be specific.**

You can address over/under fitting in a random forest model by tuning various hyper-parameters. For example:
 - n_estimators: the number of trees in the forest. 
   - decreasing the number of trees can reduce overfitting
   - increasing the number of trees can reduce underfitting
 - max_depth: sets the maximum depth of each tree in the forest.
   - decreasing the depth can reduce overfitting
   - increasing the depth can reduce underfitting
 - min_samples_split: sets the minimum number of samples required to split an internal node. 
   - increasing this parameter can reduce overfitting
   - decreasing this parameter can reduce underfitting
 - min_samples_leaf: sets the minimum number of samples required to be at a leaf node.
    - increasing this parameter can reduce overfitting
    - decreasing this parameter can reduce underfitting

**17. Consider the Customers.csv data file. This file contains basic information on customer data. The owner of a shop gets information about Customers through membership cards. This dataset consists of 2000 records and 8 columns:**
 - Customer ID: customer ID.
 - Gender: customer's gender.
 - AGE: customer's age in years.
 - Annual Income: customer's annual income.
 - Spending Score: Score assigned by the shop, based on customer behavior and spending nature.
 - Profession: customer's profession.
 - Work Experience: customer's work experience in years.
 - Family Size: number of family members.
 
**The goal is to predict Spending Score; that is, this a regression task. This is a continuation from Exercise 16 in Exam 1. In Python, answer the following:**

In [1]:
pip install optuna

Looking in indexes: https://pypi.org/simple, https://pip.repos.neuron.amazonaws.com
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install xgboost

Looking in indexes: https://pypi.org/simple, https://pip.repos.neuron.amazonaws.com
Note: you may need to restart the kernel to use updated packages.


*(a) (4 points) Using the pandas library, read the Customers.csv data file and create a dataframe called customers.*

In [7]:
import boto3
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import optuna

from sklearn.model_selection import train_test_split, KFold
from sklearn.tree import DecisionTreeRegressor, plot_tree
from sklearn.feature_selection import RFE, RFECV
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error

#defining the s3 bucket
s3 = boto3.resource('s3')
bucket_name = 'craig-shaffer-data-445-bucket'
bucket = s3.Bucket(bucket_name)

#defining the file to be read from s3 bucket
file_key = 'Customers_Exam2.csv'

bucket_object = bucket.Object(file_key)
file_object = bucket_object.get()
file_content_stream = file_object.get('Body')

#reading the datafile
customers = pd.read_csv(file_content_stream)
customers.head()

CustomerID  Gender  Age  Annual Income ($)  Spending Score (1-100)  \
0           1    Male   19              15000                      39   
1           2    Male   21              35000                      81   
2           3  Female   20              86000                       6   
3           4  Female   23              59000                      77   
4           5  Female   31              38000                      40   

      Profession  Work Experience  Family Size  
0     Healthcare                1            4  
1       Engineer                3            3  
2       Engineer                1            1  
3         Lawyer                0            2  
4  Entertainment                2            6

*(b) (3 points) Drop the Customer ID variable.*

In [8]:
#dropping CustomerID
customers = customers.drop(columns = 'CustomerID')
customers.head()

Gender  Age  Annual Income ($)  Spending Score (1-100)     Profession  \
0    Male   19              15000                      39     Healthcare   
1    Male   21              35000                      81       Engineer   
2  Female   20              86000                       6       Engineer   
3  Female   23              59000                      77         Lawyer   
4  Female   31              38000                      40  Entertainment   

   Work Experience  Family Size  
0                1            4  
1                3            3  
2                1            1  
3                0            2  
4                2            6

In [9]:
#imputing missing variables with most frequent label like in Exam 1
 
#identifying categorical variables
categorical_vars = customers.select_dtypes(include = 'object').columns.to_list()
print(categorical_vars)

#frequent values for each categorical variable (profession is the one we care about)
frequent_values = customers[categorical_vars].mode().iloc[0].to_dict()
print(frequent_values)

#creating indicator variable for observations with imputed profession (1 for imputed, 0 if not)
customers['Profession_imputed'] = customers['Profession'].isna().astype(int)

#filling missing values with most frequent labels
customers = customers.fillna(value = frequent_values)


['Gender', 'Profession']
{'Gender': 'Female', 'Profession': 'Artist'}


In [5]:
#changing Profession to dummies
customers = pd.concat([customers.drop(columns=['Profession'],axis=1), pd.get_dummies(customers[['Profession']])],axis=1)

#changing gender to labels (female-0, male-1)
customers['Gender'].replace(['Male', 'Female'], [1,0], inplace = True)

In [10]:
#adding engineered features from Exam 1

#Features 1 & 2: heredity principle features based on feature importances 
customers['heredity_feature_1'] = customers['Annual Income ($)'] * customers['Age']
customers['heredity_feature_2'] = customers['Annual Income ($)'] * customers['Work Experience']

#Feature 3: Annual Income divided by Family Size
customers['IncomeOverFamily'] = customers['Annual Income ($)'] / customers['Family Size']

#Feature 4: Work Experience ^ 2
customers['squared_WorkExperience']= np.power(customers['Work Experience'], 2)

#Feature 5: Annual Income ^ 2
customers['squared_AnnualIncome']= np.power(customers['Annual Income ($)'], 2)

#Feature 6: natural log of Family Size
customers['log_FamilySize']= np.log(customers['Family Size'])

#Feature 7 & 8: tree interactions based on patterns in Decision Tree models
customers['tree_interaction_1'] = np.where((customers['heredity_feature_1'] <=14537703.5) & 
                                         (customers['Age'] > 32.5) &
                                         (customers['heredity_feature_1'] > 3194215.0), 1, 0)
customers['tree_interaction_2'] = np.where((customers['heredity_feature_1'] <=14537703.5) & 
                                         (customers['Age'] <= 32.5) &
                                         (customers['IncomeOverFamily'] > 37307.033), 1, 0)

*(c) (5 points) Split the customers data-frame into: train (80%) and test (20%).*

In [11]:
#defining input and target
x=customers[['heredity_feature_1','IncomeOverFamily','heredity_feature_2','Annual Income ($)','Age','Work Experience','log_FamilySize']]
y=customers['Spending Score (1-100)']

#splitting the data
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,stratify=y)

*(d) (20 points) Using results from Exercise 16 part (e) in Exam 1, use the top 7 variables, build a model on the train data-frame, and tune this model using the Optuna framework. Notice that you need to tune the model based on the root mean squared error. Then, use the optimal model to make predictions on the test data-frame. Finally, compute the root mean squared error of the model. Make sure to run at least 50 trials the Optuna framework.*

In [18]:
#Model: Random Forest

SEED = 42
N_TRIALS = 50

#Optuna w/ top 7 most important features:----------
class Objective:
    
    def __init__(self, seed):
        
        self.seed = seed
        
    def __call__(self, trial):
        
        params = dict(n_estimators = trial.suggest_int('n_estimators', 100, 2000), 
                      min_samples_split = trial.suggest_int('min_samples_split', 5, 30),
                      min_samples_leaf = trial.suggest_int('min_samples_leaf', 5, 30),
                      max_depth = trial.suggest_int('max_depth', 2, 10))
        
        scores = list()
        
        skf = KFold(n_splits = 3, shuffle = True, random_state = self.seed)
        
        for train_idx, valid_idx in skf.split(x_train, y_train):
            
            x_train_1, x_valid_1 = x_train.iloc[train_idx], x_train.iloc[valid_idx]
            y_train_1, y_valid_1 = y_train.iloc[train_idx], y_train.iloc[valid_idx]
            
            rf_md = RandomForestRegressor(**params).fit(x_train_1,y_train_1)
            
            pred_valid = rf_md.predict(x_valid_1)
            
            scores.append(mean_squared_error(y_valid_1, pred_valid, squared = False))   
            
        return np.mean(scores)

study_1 = optuna.create_study(direction = 'minimize')
study_1.optimize(Objective(SEED), n_trials = N_TRIALS)

print('Best hyper-parameter combination for RandomForestRegressor with top-7 variables: \n', study_1.best_trial.params)

[I 2023-04-04 21:17:03,788] A new study created in memory with name: no-name-4dfa2078-3ce6-4b3c-a483-2c86a89e68dc
[I 2023-04-04 21:17:13,638] Trial 0 finished with value: 28.316518846655928 and parameters: {'n_estimators': 1213, 'min_samples_split': 28, 'min_samples_leaf': 29, 'max_depth': 7}. Best is trial 0 with value: 28.316518846655928.
[I 2023-04-04 21:17:17,931] Trial 1 finished with value: 28.32958270330349 and parameters: {'n_estimators': 513, 'min_samples_split': 18, 'min_samples_leaf': 17, 'max_depth': 6}. Best is trial 0 with value: 28.316518846655928.
[I 2023-04-04 21:17:19,744] Trial 2 finished with value: 28.22993100623749 and parameters: {'n_estimators': 225, 'min_samples_split': 8, 'min_samples_leaf': 6, 'max_depth': 5}. Best is trial 2 with value: 28.22993100623749.
[I 2023-04-04 21:17:33,174] Trial 3 finished with value: 28.441668363904785 and parameters: {'n_estimators': 1558, 'min_samples_split': 7, 'min_samples_leaf': 22, 'max_depth': 9}. Best is trial 2 with value

Best hyper-parameter combination for RandomForestRegressor with top-7 variables: 
 {'n_estimators': 696, 'min_samples_split': 5, 'min_samples_leaf': 12, 'max_depth': 2}


In [19]:
#running model with best parameters
rf_md = RandomForestRegressor(**study_1.best_trial.params).fit(x_train, y_train)

#predicting on test
rf_pred = rf_md.predict(x_test)

#computing rmse
rf_results = mean_squared_error(y_test, rf_pred, squared = False)
print('the RMSE of the Random Forest model is: ', rf_results)

the RMSE of the Random Forest model is:  27.734948112775353


*(e) (20 points) Using results from Exercise 16 part (e) in Exam 1, use the top 7 variables, build another model (different from the one in part (d)) on the train data-frame, and tune this model using the Optuna framework. Notice that you need to tune the model based on the root mean squared error. Then, use the optimal model to make predictions on the test data-frame. Finally, compute the root mean squared error of the model. Make sure to run at least 50 trials the Optuna framework.*

In [24]:
#Model: Gradient Boosting

SEED = 42
N_TRIALS = 50

#Optuna w/ top 7 most important features:----------
class Objective:
    
    def __init__(self, seed):
        
        self.seed = seed
        
    def __call__(self, trial):
        
        params = dict(n_estimators = trial.suggest_int('n_estimators', 100, 2000), 
                      min_samples_split = trial.suggest_int('min_samples_split', 5, 30),
                      min_samples_leaf = trial.suggest_int('min_samples_leaf', 5, 30),
                      max_depth = trial.suggest_int('max_depth', 2, 10),
                      learning_rate = trial.suggest_float('learning_rate', 0.001, .1))
        
        scores = list()
        
        skf = KFold(n_splits = 3, shuffle = True, random_state = self.seed)
        
        for train_idx, valid_idx in skf.split(x_train, y_train):
            
            x_train_1, x_valid_1 = x_train.iloc[train_idx], x_train.iloc[valid_idx]
            y_train_1, y_valid_1 = y_train.iloc[train_idx], y_train.iloc[valid_idx]
            
            gb_md = GradientBoostingRegressor(**params).fit(x_train_1,y_train_1)
            
            pred_valid = gb_md.predict(x_valid_1)
            
            scores.append(mean_squared_error(y_valid_1, pred_valid, squared = False))   
            
        return np.mean(scores)

study_2 = optuna.create_study(direction = 'minimize')
study_2.optimize(Objective(SEED), n_trials = N_TRIALS)

print('Best hyper-parameter combination for GradientBoostingRegressor with top-7 variables: \n', study_2.best_trial.params)

[I 2023-04-04 21:44:28,645] A new study created in memory with name: no-name-c567061b-8069-4621-9beb-2518a6def915
[I 2023-04-04 21:44:40,583] Trial 0 finished with value: 33.2560865757852 and parameters: {'n_estimators': 990, 'min_samples_split': 15, 'min_samples_leaf': 5, 'max_depth': 7, 'learning_rate': 0.08214941584163381}. Best is trial 0 with value: 33.2560865757852.
[I 2023-04-04 21:44:49,486] Trial 1 finished with value: 30.67711880698395 and parameters: {'n_estimators': 627, 'min_samples_split': 11, 'min_samples_leaf': 10, 'max_depth': 9, 'learning_rate': 0.016369061213959972}. Best is trial 1 with value: 30.67711880698395.
[I 2023-04-04 21:44:54,876] Trial 2 finished with value: 30.914532691824004 and parameters: {'n_estimators': 614, 'min_samples_split': 26, 'min_samples_leaf': 25, 'max_depth': 5, 'learning_rate': 0.049592992308554046}. Best is trial 1 with value: 30.67711880698395.
[I 2023-04-04 21:45:18,599] Trial 3 finished with value: 34.437088047461664 and parameters: {'

Best hyper-parameter combination for GradientBoostingRegressor with top-7 variables: 
 {'n_estimators': 340, 'min_samples_split': 8, 'min_samples_leaf': 20, 'max_depth': 2, 'learning_rate': 0.0010918677630614296}


In [26]:
#running model with best parameters
gb_md = GradientBoostingRegressor(**study_2.best_trial.params).fit(x_train, y_train)

#predicting on test
gb_pred = gb_md.predict(x_test)

#computing rmse
gb_results = mean_squared_error(y_test, gb_pred, squared = False)
print('the RMSE of the Gradient Boosting model is: ', gb_results)

the RMSE of the Gradient Boosting model is:  27.83856771994153


*(f) (20 points) Using results from Exercise 16 part (e) in Exam 1, use the top 7 variables, build another model (different from the one in parts (d)-(e)) on the train data-frame, and tune this model using the Optuna framework. Notice that you need to tune the model based on the root mean squared error. Then, use the optimal model to make predictions on the test data-frame. Finally, compute the root mean squared error of the model. Make sure to run at least 50 trials the Optuna framework.*

In [31]:
#Model: XGBoost

SEED = 42
N_TRIALS = 50

#Optuna w/ top 7 most important features:----------
class Objective:
    
    def __init__(self, seed):
        
        self.seed = seed
        
    def __call__(self, trial):
        
        params = dict(n_estimators = trial.suggest_int('n_estimators', 100, 2000),
                      max_depth = trial.suggest_int('max_depth', 2, 10),
                      min_child_weight = trial.suggest_int('min_child_weight', 2, 20),
                      learning_rate = trial.suggest_float('learning_rate', 0.001, 0.1, log = True),
                      gamma = trial.suggest_float('gamma', 0.0, 10.0),
                      colsample_bytree = trial.suggest_float('colsample_bytree', 0.2, 0.9),
                      subsample = trial.suggest_float('subsample', 0.2, 0.9))
        
        scores = list()
        
        skf = KFold(n_splits = 3, shuffle = True, random_state = self.seed)
        
        for train_idx, valid_idx in skf.split(x_train, y_train):
            
            x_train_1, x_valid_1 = x_train.iloc[train_idx], x_train.iloc[valid_idx]
            y_train_1, y_valid_1 = y_train.iloc[train_idx], y_train.iloc[valid_idx]
            
            xgb_md = XGBRegressor(**params).fit(x_train_1,y_train_1)
            
            pred_valid = xgb_md.predict(x_valid_1)
            
            scores.append(mean_squared_error(y_valid_1, pred_valid, squared = False))   
            
        return np.mean(scores)

study_3 = optuna.create_study(direction = 'minimize')
study_3.optimize(Objective(SEED), n_trials = N_TRIALS)

print('Best hyper-parameter combination for XGBRegressor with top-7 variables: \n', study_3.best_trial.params)

[I 2023-04-04 22:29:44,770] A new study created in memory with name: no-name-f4ed9ec1-aa6c-48e7-b65b-6693b95926bc
[I 2023-04-04 22:29:50,084] Trial 0 finished with value: 30.088212718420493 and parameters: {'n_estimators': 1649, 'max_depth': 10, 'min_child_weight': 20, 'learning_rate': 0.004585324232038347, 'gamma': 8.42258359764002, 'colsample_bytree': 0.8758062557586248, 'subsample': 0.698446905426491}. Best is trial 0 with value: 30.088212718420493.
[I 2023-04-04 22:29:52,954] Trial 1 finished with value: 32.230415352196694 and parameters: {'n_estimators': 1883, 'max_depth': 6, 'min_child_weight': 12, 'learning_rate': 0.03443410107272345, 'gamma': 6.598857673416909, 'colsample_bytree': 0.27204193963143014, 'subsample': 0.4209835343765423}. Best is trial 0 with value: 30.088212718420493.
[I 2023-04-04 22:29:56,073] Trial 2 finished with value: 33.03839226692882 and parameters: {'n_estimators': 1767, 'max_depth': 7, 'min_child_weight': 16, 'learning_rate': 0.041268157039171395, 'gamma

Best hyper-parameter combination for XGBRegressor with top-7 variables: 
 {'n_estimators': 1517, 'max_depth': 2, 'min_child_weight': 13, 'learning_rate': 0.0027160167543365107, 'gamma': 7.714742420788344, 'colsample_bytree': 0.39086992070842974, 'subsample': 0.3693192002420051}


In [32]:
#running model with best parameters
xgb_md = XGBRegressor(**study_3.best_trial.params).fit(x_train, y_train)

#predicting on test
xgb_pred = xgb_md.predict(x_test)

#computing rmse
xgb_results = mean_squared_error(y_test, xgb_pred, squared = False)
print('the RMSE of the Gradient Boosting model is: ', xgb_results)

the RMSE of the Gradient Boosting model is:  27.837240813963206


*(g) (20 points) Using the predictions from parts (d), (e), and (f), ensemble them by taking the average of those predictions. Finally, compute the root mean squared error of this ensemble.*

In [36]:
ensemble_pred = (rf_pred + gb_pred + xgb_pred)/3
ensemble_results = mean_squared_error(y_test, ensemble_pred, squared = False)
print('RMSE of the ensemble: ', ensemble_results)

RMSE of the ensemble:  27.78771261215174


*(h) (5 points), Using the results from parts (d) to (g), what model/approach would use to predict Spending Score. Be specific.*

In [38]:
print('RMSE of Random Forest: ', rf_results)
print('RMSE of Gradient Boosting: ', gb_results)
print('RMSE of XGBoost: ', xgb_results)
print('RMSE of the ensemble: ', ensemble_results)

RMSE of Random Forest:  27.734948112775353
RMSE of Gradient Boosting:  27.83856771994153
RMSE of XGBoost:  27.837240813963206
RMSE of the ensemble:  27.78771261215174


Based on the results above, the random forest model is better for predicting Spending Score because it has the lowest RMSE (27.73).